### PR TITLE
feat(sdk): visual annotations of resources

### PR DESCRIPTION
**Pull request by polamoros**
*By submitting this pull request, I confirm that my contribution is made under the terms of the 
[Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingsdk/src/core/tree.ts [hunk 1]
```diff
@@ -134,6 +161,17 @@ function synthAttributes(
   return undefined;
 }
 
+function synthDisplay(construct: IConstruct): DisplayInfo | undefined {
+  if (construct instanceof Resource) {
```
Comment (eladb) [inline libs/wingsdk/src/core/tree.ts:165]: Why is this needed?

File: libs/wingsdk/src/core/tree.ts [hunk 2]
```diff
@@ -33,6 +34,31 @@ export interface ConstructTreeNode {
...
+
+  /**
+   * Description of the resource.
+   */
+  readonly description: string;
+
+  /**
+   * Whether the resource should be hidden from the UI.
+   */
+  readonly hidden: boolean;
```
Comment (eladb) [inline libs/wingsdk/src/core/tree.ts:61]:
```suggestion
  /**
   * Title of the resource.
   * @default - The type and/or identifier of the resource
   */
  readonly title?: string;

  /**
   * Description of the resource.
   * @default - No description
   */
  readonly description?: string;

  /**
   * Whether the resource should be hidden from the UI.
   * @default false (visible)
   */
  readonly hidden?: boolean;
```

File: libs/wingsdk/src/core/resource.ts [hunk 3]
```diff
@@ -229,3 +234,33 @@ export interface OperationAnnotation {
...
 }
+
+/**
+ * Information on how to display a resource in the UI.
+ */
+export class Display {
+  /**
+   * Title of the resource.
+   */
+  public title: string;
```
Comment (eladb) [inline libs/wingsdk/src/core/resource.ts:245]: All of these should be optional (`?`).

File: libs/wingsdk/src/core/resource.ts [hunk 4]
```diff
@@ -229,3 +234,33 @@ export interface OperationAnnotation {
...
+
+  /**
+   * Whether the resource should be hidden from the UI.
+   */
+  public hidden: boolean;
+
+  public constructor(
+    title: string = "",
+    description: string = "",
+    hidden: boolean = false
```
Comment (eladb) [inline libs/wingsdk/src/core/resource.ts:260]:
```suggestion
    title?: string,
    description?: string,
    hidden?: boolean
```

File: libs/wingsdk/test/target-sim/__snapshots__/bucket.test.ts.snap [hunk 5]
```diff
@@ -33,6 +33,11 @@ Object {
             "fqn": "constructs.Construct",
             "version": "10.0.25",
           },
+          "display": Object {
```
Comment (eladb) [inline libs/wingsdk/test/target-sim/__snapshots__/bucket.test.ts.snap:36]: Omit `display` if if's all defaults (`undefined`).

**Review by eladb — CHANGES_REQUESTED** [review]
@polamoros This looks awesome!

A few notes:
1. Add some unit tests for `display`.
2. Make the `Inflight` resource `hidden: true`.
3. Make the `Logger` resource `hidden: true`.

File: libs/wingsdk/src/core/tree.ts [hunk 6]
```diff
@@ -33,6 +34,34 @@ export interface ConstructTreeNode {
    * Information on the construct class that led to this node, if available.
    */
   readonly constructInfo?: ConstructInfo;
+
+  /**
+   * Information on how to display this node in the UI.
+   */
+  readonly display?: DisplayInfo;
```
Comment (Chriscbr) [inline libs/wingsdk/src/core/tree.ts:41]: Instead of adding a new field here, could we add this information as another attribute (like the existing connections and stateful attributes)? This can be done by updating the `_inspect` method:

https://github.com/winglang/wing/blob/aa8e4510cd508ed38213d0afda65d6f4cc860cd1/libs/wingsdk/src/core/resource.ts#L196-L208

(sorry in advance, this needs to be better documented :))

Comment (eladb) [reply libs/wingsdk/src/core/tree.ts:41]: @Chriscbr I am curious, what's the value you see in using these attributes instead of simply extending the base schema?

Comment (Chriscbr) [reply libs/wingsdk/src/core/tree.ts:41]: @eladb I guess there isn't much of a difference - my thought process back in the day was "it looks like tree.json in the CDK puts most optional metadata in this `attributes` field, so if we want to play well, we can put stuff there too." There's an also an argument that by putting it in this `attributes` field, it signals more clearly that the attribute isn't required / isn't guaranteed to exist. I'm ok with the current place if you feel it makes more sense though 👍

Comment (eladb) [reply libs/wingsdk/src/core/tree.ts:41]: I feel there's value in lifting up everything into the schema as first class citizens and getting rid of "attributes" as a 2nd class protocol.

File: libs/wingsdk/src/cloud/logger.ts [hunk 7]
```diff
@@ -12,6 +12,11 @@ export const LOGGER_SYMBOL = Symbol.for(LOGGER_TYPE);
 export abstract class LoggerBase extends Resource {
   public readonly stateful = true;
 
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+    this.display.hidden = true;
```
Comment (eladb) [inline libs/wingsdk/src/cloud/logger.ts:17]: Maybe add a description as well?

File: libs/wingsdk/src/core/resource.ts [hunk 8]
```diff
@@ -294,3 +304,52 @@ export interface OperationAnnotation {
...
+
+  /**
+   * Whether the resource should be hidden from the UI.
+   */
+  public hidden?: boolean;
+
+  public constructor(props?: DisplayProps) {
+    this.title = props?.title;
+    this.description = props?.description;
+    this.hidden = props?.hidden;
```
Comment (eladb) [inline libs/wingsdk/src/core/resource.ts:353]:
```suggestion
  public constructor(props: DisplayProps = {}) {
    this.title = props.title;
    this.description = props.description;
    this.hidden = props.hidden;
```

File: libs/wingsdk/src/core/tree.ts [hunk 9]
```diff
@@ -134,6 +164,17 @@ function synthAttributes(
   return undefined;
 }
 
+function synthDisplay(construct: IConstruct): DisplayInfo | undefined {
+  if (
+    construct instanceof Resource &&
```
Comment (eladb) [inline libs/wingsdk/src/core/tree.ts:169]: This won't work for the Inflight class

**Review by eladb — APPROVED** [review]

File: libs/wingsdk/src/core/tree.ts (hunk 9 above)
Comment (polamoros) [reply libs/wingsdk/src/core/tree.ts]: fixed!

**Comment by polamoros** [comment]
It seems `package-lock.json`s are not updated on main

File: libs/wingsdk/src/cloud/logger.ts (hunk 7 above)
Comment (eladb) [reply libs/wingsdk/src/cloud/logger.ts:21]: LOL that's not what I meant, I meant `this.display.description`

**Review by eladb — COMMENTED** [review]
@Chriscbr can you take a look?

File: libs/wingsdk/src/core/resource.ts [hunk 10]
```diff
@@ -294,3 +304,52 @@ export interface OperationAnnotation {
...
+
+  /**
+   * Description of the resource.
+   */
+  public description?: string;
+
+  /**
+   * Whether the resource should be hidden from the UI.
+   */
+  public hidden: boolean;
```
[the remark this answers is not in the document]
Comment (polamoros) [reply libs/wingsdk/src/core/resource.ts]: done 🙏

File: libs/wingsdk/test/target-sim/logger.test.ts [hunk 11]
```diff
@@ -33,3 +33,23 @@ test("inflight uses a logger", async () => {
   expect(listMessages(s)).toMatchSnapshot();
   expect(app.snapshot()).toMatchSnapshot();
 });
+
+test("Logger has display hidden property set to true", async () => {
```
Comment (Chriscbr) [inline libs/wingsdk/test/target-sim/logger.test.ts:37]: Could we also add a test somewhere that built-in resources (Bucket, Counter, Function) are not hidden by default?

File: libs/wingsdk/src/cloud/logger.ts (hunk 7 above)
Comment (polamoros) [reply libs/wingsdk/src/cloud/logger.ts:21]: Why a `display.description` of a Resource that will not be displayed on the Console?

Comment (eladb) [reply libs/wingsdk/src/cloud/logger.ts:21]: Mmmm. Good question... The way I see it, we should include a description for all the resources in the SDK, whether they are displayed or not. The reason:

1. The description can be used in other places, not just the console.
2. We might add support in the console to show hidden nodes at some point, in which case it makes sense to have a description for those nodes.

But that's not really critical... We can also add that later (and perhaps better to add it to all resources at once).

Comment (polamoros) [reply libs/wingsdk/src/cloud/logger.ts:21]: Agree on both points 👍 

Let me add a title and a description to all resources on this PR.

Comment (eladb) [reply libs/wingsdk/src/cloud/logger.ts:21]: 👍

Comment (polamoros) [reply libs/wingsdk/src/cloud/logger.ts:18]: I added titles and descriptions from the [SDK ref](https://docs.winglang.io/reference/sdk).. Not the best ones I think.. :)

File: libs/wingsdk/src/cloud/bucket.ts [hunk 12]
```diff
@@ -25,6 +25,10 @@ export abstract class BucketBase extends Resource {
   public readonly stateful = true;
   constructor(scope: Construct, id: string, props: BucketProps) {
     super(scope, id);
+
+    this.display.title = "Bucket";
+    this.display.description = "Represents a cloud object store";
```
Comment (eladb) [inline libs/wingsdk/src/cloud/bucket.ts:30]:
```suggestion
    this.display.description = "A cloud object store";
```

File: libs/wingsdk/src/cloud/counter.ts [hunk 13]
```diff
@@ -31,6 +31,10 @@ export abstract class CounterBase extends Resource {
 
   constructor(scope: Construct, id: string, props: CounterProps = {}) {
     super(scope, id);
+
+    this.display.title = "Counter";
+    this.display.description = "Represents a distributed atomic counter";
```
Comment (eladb) [inline libs/wingsdk/src/cloud/counter.ts:36]:
```suggestion
    this.display.description = "A distributed atomic counter";
```

File: libs/wingsdk/src/cloud/function.ts [hunk 14]
```diff
@@ -46,6 +46,9 @@ export abstract class FunctionBase extends Resource implements IInflightHost {
   ) {
     super(scope, id);
 
+    this.display.title = "Function";
+    this.display.description = "Represents a function";
```
Comment (eladb) [inline libs/wingsdk/src/cloud/function.ts:50]:
```suggestion
    this.display.description = "A cloud function (FaaS)";
```

File: libs/wingsdk/src/cloud/queue.ts [hunk 15]
```diff
@@ -33,6 +33,10 @@ export abstract class QueueBase extends Resource {
   public readonly stateful = true;
   constructor(scope: Construct, id: string, props: QueueProps = {}) {
     super(scope, id);
+
+    this.display.title = "Queue";
+    this.display.description = "Represents a queue";
```
Comment (eladb) [inline libs/wingsdk/src/cloud/queue.ts:38]:
```suggestion
    this.display.description = "A distributed message queue";
```

File: libs/wingsdk/src/cloud/topic.ts [hunk 16]
```diff
@@ -17,6 +17,10 @@ export abstract class TopicBase extends Resource {
   public readonly stateful = true;
   constructor(scope: Construct, id: string, props: TopicProps = {}) {
     super(scope, id);
+
+    this.display.title = "Topic";
+    this.display.description = "Represents a topic";
```
Comment (eladb) [inline libs/wingsdk/src/cloud/topic.ts:22]:
```suggestion
    this.display.description = "A pub/sub notification topic";
```

File: libs/wingsdk/src/core/tree.ts [hunk 17]
```diff
@@ -134,6 +165,21 @@ function synthAttributes(
   return undefined;
 }
 
+function isIResource(construct: IConstruct): construct is IResource {
+  return construct instanceof Resource || construct instanceof Inflight;
```
Comment (eladb) [inline libs/wingsdk/src/core/tree.ts:169]: Meh... @Chriscbr what do you think about this? Not ideal... but I guess `Inflight` is going away anyway, right?

**Review by eladb — APPROVED** [review]

File: libs/wingsdk/src/core/tree.ts (hunk 17 above)
Comment (Chriscbr) [reply libs/wingsdk/src/core/tree.ts:169]: Yeah, this looks okay to me. 🙂 BTW we could also opt for duck typing:

```ts
function isDisplayable(construct: IConstruct): boolean {
  if (construct.display) {
    return true;
  }
  return false;
}
```

Comment (skyrpex) [reply libs/wingsdk/src/core/tree.ts:169]: ^ That would need either ts-ignore or a cast, right?

Comment (polamoros) [reply libs/wingsdk/src/core/tree.ts:169]: I think we can let it as it is. What do you think?

**Review by eladb — APPROVED** [review]
🎉

File: libs/wingsdk/src/core/tree.ts (hunk 17 above)
Comment (Chriscbr) [reply libs/wingsdk/src/core/tree.ts:169]: Sounds good
